ktor-client/ktor-client-apache/jvm/src/io/ktor/client/engine/apache/ContentReader.kt

```diff
@@ -0,0 +1,119 @@
+/*
+ * Copyright 2014-2019 JetBrains s.r.o and contributors. Use of this source code is governed by the Apache 2.0 license.
+ */
+
+package io.ktor.client.engine.apache
+
+import kotlinx.coroutines.*
+import kotlinx.coroutines.io.*
+import org.apache.http.nio.*
+import java.nio.channels.*
+import java.util.concurrent.*
+import kotlin.coroutines.*
+import kotlin.coroutines.intrinsics.*
+
+internal class ContentReader : CoroutineDispatcher(), Continuation<Unit> {
+    private lateinit var currentDecoder: ContentDecoder
+    private var control: IOControl? = null
+
+    private val state: CompletableJob = Job()
+    private var hasData = true
+
+    private val _content = ByteChannel()
+
+    val content: ByteReadChannel get() = _content
+
+    private val events = ConcurrentLinkedQueue<Runnable>()
+
+    override val context: CoroutineContext = this
+
+    private val handler = suspend handler@{
+        try {
+            yield()
+            if (!state.isActive) return@handler
```

Comment:
Is it ok to not close `_content` if cancelled immediately?

Reply:
Is it ok to keep hasData = true in this case?

Reply:
We close `_content` and switch `hasData` in finally block here. It should be enough.